### ISSUE TITLE
Remove unused array elements in QScript

### ISSUE THREAD

**Issue by alissapajer** [quasar-analytics/quasar#1745]
QScript stores provenance information in arrays. Sometimes this provenance information ultimately goes unused, meaning it can be removed from the QScript.

In the following example, only indices 2 and 3 are dereferenced from the array, meaning the array can be rewritten and the derefs can be reindexed.
```scala
    "convert a filter" in {
      // "select * from foo where bar between 1 and 10"
      convert(
        listContents.some,
        StdLib.set.Filter(
          lpRead("/bar"),
          relations.Between(
            structural.ObjectProject(lpRead("/bar"), lpf.constant(Data.Str("baz"))).embed,
            lpf.constant(Data.Int(1)),
            lpf.constant(Data.Int(10))).embed).embed) must
      beSome(beQScript(chain(
        ReadR(rootDir </> file("bar")),
        QC.inj(LeftShift((),
          HoleF,
          IncludeId,
          Free.roll(ConcatArrays(
            Free.roll(ConcatArrays(
              Free.roll(ConcatArrays(
                Free.roll(MakeArray(ProjectIndexR(RightSideF, IntLit(0)))),
                // FIXME: #1622
                Free.roll(MakeArray(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(RightSideF, IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)))))),
              Free.roll(MakeArray(ProjectIndexR(RightSideF, IntLit(1)))))),
            Free.roll(MakeArray(
              Free.roll(Between(
                ProjectFieldR(
                  ProjectIndexR(RightSideF, IntLit(1)),
                  StrLit("baz")),
                IntLit(1),
                IntLit(10))))))))),
        QC.inj(Filter((), ProjectIndexR(HoleF, IntLit(3)))),
QC.inj(Map((), ProjectIndexR(HoleF, IntLit(2)))))))
```

**Comment by alissapajer** [comment]
This relates to #1622 in that sometimes we don't reference the excessive multiple array projections reported there, and so they can be pruned in this way. But when they cannot be pruned, that bug will still be hit.

**Comment by alissapajer** [comment]
Closed by merge of #1747
